pkg/logs/input/kubernetes/launcher.go

```diff
@@ -0,0 +1,219 @@
+	sources                   *config.LogSources
+	sourcesByContainer        map[string]*config.LogSource
+	stopped                   chan struct{}
+	kubeutil                  *kubelet.KubeUtil
+	dockerSources             chan *config.LogSource
+	dockerAddedServices       chan *service.Service
+	dockerRemovedServices     chan *service.Service
+	containerdSources         chan *config.LogSource
+	containerdAddedServices   chan *service.Service
+	containerdRemovedServices chan *service.Service
+}
+
+// NewLauncher returns a new launcher.
+func NewLauncher(sources *config.LogSources, services *service.Services) (*Launcher, error) {
+	kubeutil, err := kubelet.GetKubeUtil()
+	if err != nil {
+		return nil, err
+	}
+	launcher := &Launcher{
+		sources:                   sources,
+		sourcesByContainer:        make(map[string]*config.LogSource),
+		stopped:                   make(chan struct{}),
+		kubeutil:                  kubeutil,
+		dockerSources:             sources.GetSourceStreamForType(config.DockerType),
+		dockerAddedServices:       services.GetAddedServices(service.Docker),
+		dockerRemovedServices:     services.GetRemovedServices(service.Docker),
+		containerdSources:         sources.GetSourceStreamForType(config.ContainerdType),
+		containerdAddedServices:   services.GetAddedServices(service.Containerd),
+		containerdRemovedServices: services.GetRemovedServices(service.Containerd),
+	}
+	err = launcher.setup()
+	if err != nil {
+		return nil, err
+	}
+	return launcher, nil
+}
+
+// setup initializes the pod watcher and the tagger.
+func (l *Launcher) setup() error {
+	var err error
+	// initialize the tagger to collect container tags
+	err = tagger.Init()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Start starts the launcher
+func (l *Launcher) Start() {
+	log.Info("Starting Kubernetes launcher")
+	go l.run()
+}
+
+// Stop stops the launcher
+func (l *Launcher) Stop() {
+	log.Info("Stopping Kubernetes launcher")
+	l.stopped <- struct{}{}
+}
+
+// run handles new and deleted pods,
+// the kubernetes launcher consumes new and deleted services pushed by the autodiscovery
+func (l *Launcher) run() {
+	for {
+		select {
+		case service := <-l.dockerAddedServices:
+			l.addSources(service)
+		case service := <-l.dockerRemovedServices:
+			l.removeSources(service)
+		case <-l.dockerSources:
+			// The annotation are resolved through the pod object. We don't need
+			// to process the source here
+			continue
+		case service := <-l.containerdAddedServices:
+			l.addSources(service)
+		case service := <-l.containerdRemovedServices:
+			l.removeSources(service)
+		case <-l.containerdSources:
+			// The annotation are resolved through the pod object. We don't need
+			// to process the source here
+			continue
+		case <-l.stopped:
+			return
+		}
+	}
+}
+
+// addSources creates a new log-source from a service by resolving the
+// pod linked to the entityID of the service
+func (l *Launcher) addSources(service *service.Service) {
```

Comment:
Should this method be renamed into `addSource` and just create a new file source for the `service.Identifier` ? This would enable us to get rid of the map but we'd have to navigate in the pod structure multiple times, dunno what's the best solution. WDYT ?

Reply:
I think it makes no difference: 
Here, you get a container, you add a source for all container in its pod, then you receive the other containers of the same pod and you do nothing

Or you get a container, you have to find the matching one in the pod and add the source

I think the current logic is simpler since you don't have to handle edge cases

Reply:
You're right the behaviour does not change but this method is meant to create a new source for a given service, here it's not what it does, instead it maps one service to a set of sources which is not semantically correct IMO. I'd rather make it clear mapping one service to one source so that we can get rid of the map.